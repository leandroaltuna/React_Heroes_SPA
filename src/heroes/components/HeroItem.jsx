import { Link } from 'react-router-dom';


const CharactersByHero = ({ alter_ego, characters }) => {
    if ( alter_ego === characters ) return(<></>);
    return <p>{ characters }</p>;
}


export const HeroItem = ({ 
    id,
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,   
 }) => {

    const heroImgUrl = `/heroes/${ id }.jpg`;
    // const heroImgUrl = `/assets/heroes/${ id }.jpg`;
    // const charactersByHero = ( <p>{ characters }</p> );

    return (
        <>
            <div className="col mb-1 animate__animated animate__fadeIn">
                <div className="card">

                    <div className="row no-gutters">
                        
                        <div className="col-4">
                            <img 
                                className="card-img"
                                src={ heroImgUrl } 
                                alt={ superhero } 
                            />
                        </div>

                        <div className="col-8">
                            <div className="card-body">

                                <h5 className="card-title">{ superhero }</h5>
                                <p className="card-text">{ alter_ego }</p>
                                {/* {
                                    ( alter_ego !== characters ) && charactersByHero
                                } */}
                                <CharactersByHero 
                                    alter_ego={ alter_ego } 
                                    characters={ characters } 
                                />
                                
                                <p className="card-text">
                                    <small className="text-muted">{ first_appearance }</small>
                                </p>

                                <Link to={ `/hero/${ id }` }>
                                    more...
                                </Link>

                            </div>
                        </div>

                    </div>

                </div>
            </div>
        </>
    )

}
