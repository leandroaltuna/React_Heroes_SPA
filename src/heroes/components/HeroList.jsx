import { useMemo } from 'react';
import { getHeroesByPublisher } from '../helpers';
import { HeroItem } from './HeroItem';


export const HeroList = ({ publisher }) => {

    // console.log( publisher );
    const heroes = useMemo( () => getHeroesByPublisher( publisher ), [ publisher ] );

    return (
        
        <>
            <div className="row rows-cols-1 row-cols-md-3 g-3">
                {
                    heroes.map( hero => (

                        <HeroItem
                            key={ hero.id }
                            { ...hero }
                        />
                        // <li
                        //     key={ hero.id }
                        // >
                        //     { hero.superhero }
                        // </li>

                    ))
                }
            </div>
        </>

    )

}
