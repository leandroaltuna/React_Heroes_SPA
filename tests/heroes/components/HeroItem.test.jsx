import { render, screen } from '@testing-library/react';
import { HeroItem } from '../../../src/heroes/components/HeroItem';
import { MemoryRouter } from 'react-router-dom';


describe( 'Testing <HeroItem />', () => { 

    test( 'Debe de mostrar Hero', () => { 
        
        const heroInfo = {
            'id': 'dc-batman',
            'superhero':'Batman', 
            'publisher':'DC Comics', 
            'alter_ego':'Bruce Wayne',
            'first_appearance':'Detective Comics #27',
            'characters':'Bruce Wayne'
        };

        const { container } = render(
            <MemoryRouter>
                <HeroItem { ...heroInfo } />
            </MemoryRouter>
        );
        // screen.debug();

        expect( container ).toMatchSnapshot();

        const heroImg = screen.getByRole( 'img' );
        expect( heroImg.src ).toContain( `/heroes/${ heroInfo.id }.jpg` );

        //? OJO, necesita un arial-label or name para que funcione.
        const moreLink = screen.getByRole( 'link' );
        expect( moreLink.href ).toContain( `/hero/${ heroInfo.id }` );
        
    });
    
});