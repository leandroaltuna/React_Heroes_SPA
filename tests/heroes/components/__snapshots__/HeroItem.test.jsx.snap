// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <HeroItem /> Debe de mostrar Hero 1`] = `
<div>
  <div
    class="col mb-1 animate__animated animate__fadeIn"
  >
    <div
      class="card"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="col-4"
        >
          <img
            alt="Batman"
            class="card-img"
            src="/heroes/dc-batman.jpg"
          />
        </div>
        <div
          class="col-8"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              Batman
            </h5>
            <p
              class="card-text"
            >
              Bruce Wayne
            </p>
            <p
              class="card-text"
            >
              <small
                class="text-muted"
              >
                Detective Comics #27
              </small>
            </p>
            <a
              aria-label="moreHero"
              href="/hero/dc-batman"
            >
              more...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
